### PR TITLE
replace grpc events by ra events

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

File: cloud2cloud-connector/service/deviceSubscriptionHandlers.go [hunk 1]
```diff
@@ -46,6 +46,10 @@ func (h deviceSubscriptionHandlers) CreateResource(ctx context.Context, event *e
 	return fmt.Errorf("not supported")
 }
 
+func (h deviceSubscriptionHandlers) DeviceMetadataUpdate(ctx context.Context, event *events.DeviceMetadataUpdatePending) error {
```
Comment (ondrejtomcik) [inline cloud2cloud-connector/service/deviceSubscriptionHandlers.go]: `UpdateDeviceMetadata`

File: coap-gateway/service/client.go [hunk 2]
```diff
@@ -89,15 +87,14 @@ func (a *authorizationContext) IsValid() error {
...
 //Client a setup of connection
 type Client struct {
 	server   *Service
 	coapConn *tcp.ClientConn
 
-	observedResources     map[string]map[int64]*observedResource // [deviceID][instanceID]
-	observedResourcesLock sync.Mutex
+	observedResources         map[string]map[int64]*observedResource // [deviceID][instanceID]
+	observedResourcesLock     sync.Mutex
+	observedResourcesDisabled bool
```
Comment (ondrejtomcik) [inline coap-gateway/service/client.go]: what exactly is disabled? resourceObservationDisabled?

File: coap-gateway/service/client.go [hunk 3]
```diff
@@ -849,3 +870,77 @@ func (client *Client) GetContext() (context.Context, context.CancelFunc) {
 	}
 	return kitNetGrpc.CtxWithOwner(client.Context(), authCtx.GetUserID()), func() {}
 }
+
+func (client *Client) sendErrorConfirmDeviceMetadataUpdate(userID string, event *events.DeviceMetadataUpdatePending, errToSend error) {
+	ctx, err := client.server.ServiceRequestContext(userID)
+	if err != nil {
+		log.Errorf("cannot send error via confirm resource create: %v", err)
```
Comment (ondrejtomcik) [inline coap-gateway/service/client.go]: wrong error

File: coap-gateway/service/client.go [hunk 4]
```diff
@@ -849,3 +870,77 @@ func (client *Client) GetContext() (context.Context, context.CancelFunc) {
...
+			ShadowSynchronization: event.GetShadowSynchronization(),
+		},
+		CommandMetadata: &commands.CommandMetadata{
+			ConnectionId: client.remoteAddrString(),
+			Sequence:     client.coapConn.Sequence(),
+		},
+		Status: commands.Status_UNAUTHORIZED,
+	})
+	if err != nil {
+		log.Errorf("cannot send error via confirm resource create: %v", err)
```
Comment (ondrejtomcik) [inline coap-gateway/service/client.go]: wrong error message

File: coap-gateway/service/client.go [hunk 5]
```diff
@@ -849,3 +870,77 @@ func (client *Client) GetContext() (context.Context, context.CancelFunc) {
...
+		DeviceId:      event.GetDeviceId(),
+		CorrelationId: event.GetAuditContext().GetCorrelationId(),
+		Confirm: &commands.ConfirmDeviceMetadataUpdateRequest_ShadowSynchronization{
+			ShadowSynchronization: event.GetShadowSynchronization(),
+		},
+		CommandMetadata: &commands.CommandMetadata{
+			ConnectionId: client.remoteAddrString(),
+			Sequence:     client.coapConn.Sequence(),
+		},
+		Status: commands.Status_UNAUTHORIZED,
```
Comment (ondrejtomcik) [inline coap-gateway/service/client.go]: why do you set status to unauthorized?

File: coap-gateway/service/client.go [hunk 6]
```diff
@@ -849,3 +870,77 @@ func (client *Client) GetContext() (context.Context, context.CancelFunc) {
...
+
+func (client *Client) setShadowSynchronization(ctx context.Context, deviceID string, disabled bool) {
+	client.observedResourcesLock.Lock()
+	defer client.observedResourcesLock.Unlock()
+	if disabled == client.observedResourcesDisabled {
+		return
+	}
+
+	if disabled {
+		client.observedResourcesDisabled = true
```
Comment (ondrejtomcik) [inline coap-gateway/service/client.go]: resource itself is disabled or what is really disabled?

File: coap-gateway/service/client.go [hunk 7]
```diff
@@ -849,3 +870,77 @@ func (client *Client) GetContext() (context.Context, context.CancelFunc) {
...
+		client.cleanObservedResourcesLocked()
+	} else {
+		client.observedResourcesDisabled = false
+		client.registerObservationsForPublishedResourcesLocked(ctx, deviceID)
+	}
+}
+
+func (client *Client) DeviceMetadataUpdate(ctx context.Context, event *events.DeviceMetadataUpdatePending) error {
```
Comment (ondrejtomcik) [inline coap-gateway/service/client.go]: `UpdateDeviceMetadata`

File: coap-gateway/service/client.go [hunk 8]
```diff
@@ -849,3 +870,77 @@ func (client *Client) GetContext() (context.Context, context.CancelFunc) {
...
+	}
+	if event.GetShadowSynchronization() == nil {
+		return nil
+	}
+	sendConfirmCtx, err := client.server.ServiceRequestContext(authCtx.GetUserID())
+	if err != nil {
+		return err
+	}
+
+	client.setShadowSynchronization(sendConfirmCtx, event.GetDeviceId(), event.GetShadowSynchronization().GetDisabled())
```
Comment (ondrejtomcik) [inline coap-gateway/service/client.go]: Would be more understandable to rename Getter GetDisabled to IsDisabled.

File: coap-gateway/service/resourceDiscovery.go [hunk 9]
```diff
@@ -57,35 +58,38 @@ func makeDiscoveryResp(isTLSListener bool, serverAddr string, getResourceLinksCl
...
-		if resource.Anchor == "" {
-			resource.Anchor = "ocf://" + resource.DeviceID
-		}
-		//override EndpointInformations to cloud address
-		resource.Endpoints = []schema.Endpoint{
-			{
-				URI:      ep,
-				Priority: 1,
-			},
+		links := commands.ResourcesToResourceLinks(snapshot.GetResources())
```
Comment (ondrejtomcik) [inline coap-gateway/service/resourceDiscovery.go:76]: Why you don't make `snapshot.GetResources().ToResourceLinks()` reciever function?

File: coap-gateway/service/resourceDiscovery.go [hunk 10]
```diff
@@ -57,35 +58,38 @@ func makeDiscoveryResp(isTLSListener bool, serverAddr string, getResourceLinksCl
...
+		for i := range links {
+			links[i].Href = makeHref(links[i].DeviceID, links[i].Href)
+			links[i].ID = ""
+			//set anchor if it is not set
```
Comment (ondrejtomcik) [inline coap-gateway/service/resourceDiscovery.go]: This comment has no added value.

File: docs/guide/developing/grpc-client.md [hunk 11]
```diff
@@ -46,14 +46,14 @@ The `RetrieveResourcesValues` command supports various filter options. If all of
...
+**To control what will be pushed to the stream, send a `SubscribeToEvents` message with option:**
 - `filter_by.devices_event.filter_events` set to e.g. `ONLINE` to receive **devices events** which changed their status to `ONLINE`
 - `filter_by.device_event.{device_id, filter_events}` set to e.g. `RESOURCE_PUBLISHED` to receive **device events**
 - `filter_by.device_event.{resource_id.{device_id, href}, filter_events}` set to e.g. `CONTENT_CHANGED` to receive **resource events**
 
-First event returned after the successful subscription is of type `OperationProcessed`. Property `OperationProcessed.error_status.code` contains information if the subscription was successful. If it was successful, property `subscriptionId` is set. All events belonging to single `SubscribeForEvents` request are then identified by this `subscriptionId`.
+First event returned after the successful subscription is of type `OperationProcessed`. Property `OperationProcessed.error_status.code` contains information if the subscription was successful. If it was successful, property `subscriptionId` is set. All events belonging to single `SubscribeToEvents` request are then identified by this `subscriptionId`.
```
Comment (ondrejtomcik) [inline docs/guide/developing/grpc-client.md:56]: Please update plgd.dev repository, not this one. And extend it by filters you added and possibility to specify wildcard.

File: grpc-gateway/client/getDevices_test.go [hunk 12]
```diff
@@ -15,7 +15,7 @@ import (
 
 func sortDevices(s map[string]*client.DeviceDetails) map[string]*client.DeviceDetails {
 	for key, x := range s {
-		x.Resources = test.SortResources(x.Resources)
+		x.Resources = test.CleanUpArrayResources(x.Resources)
```
Comment (ondrejtomcik) [inline grpc-gateway/client/getDevices_test.go]: ResourcesArray ...

File: grpc-gateway/client/observeDevices.go [hunk 13]
```diff
@@ -30,17 +31,17 @@ type devicesObservation struct {
...
 
-func (o *devicesObservation) HandleDeviceOnline(ctx context.Context, val *pb.Event_DeviceOnline) error {
-	return o.h.Handle(ctx, DevicesObservationEvent{
-		DeviceIDs: val.GetDeviceIds(),
-		Event:     DevicesObservationEvent_ONLINE,
-	})
-}
-
-func (o *devicesObservation) HandleDeviceOffline(ctx context.Context, val *pb.Event_DeviceOffline) error {
+func (o *devicesObservation) HandleDeviceMetadataUpdated(ctx context.Context, val *events.DeviceMetadataUpdated) error {
```
Comment (ondrejtomcik) [inline grpc-gateway/client/observeDevices.go:34]: Please don't forget to document this event and its behaviour - especially with validUntil.

File: grpc-gateway/pb/devices.proto [hunk 14]
```diff
@@ -111,23 +87,6 @@ message Event {
   message DeviceUnregistered {
     repeated string device_ids = 1;
   }
-  message DeviceOnline {
-    repeated string device_ids = 1;
-  }
-  message DeviceOffline {
-    repeated string device_ids = 1;
-  }
-  message ResourcePublished {
```
Comment (ondrejtomcik) [inline grpc-gateway/pb/devices.proto:120]: Why did you remove also ResourcePublished, ResourceUnpublished and Resource Changed? And didn't add ResourceMetadataUpdated?

File: grpc-gateway/pb/retrieveDevicesMetadata.proto [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+package ocf.cloud.grpcgateway.pb;
+
+import "github.com/plgd-dev/cloud/resource-aggregate/pb/commands.proto";
+
+
+option go_package = "github.com/plgd-dev/cloud/grpc-gateway/pb;pb";
+
+message RetrieveDevicesMetadataRequest{
+    repeated string device_ids_filter = 1;
```
Comment (ondrejtomcik) [inline grpc-gateway/pb/retrieveDevicesMetadata.proto:11]: Why you are not supporting device type filter?

File: grpc-gateway/pb/service.proto [hunk 16]
```diff
@@ -5,33 +5,43 @@ package ocf.cloud.grpcgateway.pb;
...
   // Delete resource at the device.
-  rpc DeleteResource(DeleteResourceRequest) returns (DeleteResourceResponse) {}
+  rpc DeleteResource(DeleteResourceRequest) returns (ocf.cloud.resourceaggregate.pb.ResourceDeleted) {}
 
   // Create resource at the device.
-  rpc CreateResource(CreateResourceRequest) returns (CreateResourceResponse) {}
+  rpc CreateResource(CreateResourceRequest) returns (ocf.cloud.resourceaggregate.pb.ResourceCreated) {}
+
+  // UpdateDeviceShadowSynchronization enables/disables shadow synchronization of device.
+  rpc UpdateDeviceShadowSynchronization(UpdateDeviceShadowSynchronizationRequest) returns (UpdateDeviceShadowSynchronizationResponse) {}
```
Comment (ondrejtomcik) [inline grpc-gateway/pb/service.proto:42]: Ehm, Did we decided to make a unique command for it?

File: coap-gateway/service/signIn.go [hunk 17]
```diff
@@ -57,9 +57,43 @@ func registerObservationsForPublishedResources(ctx context.Context, client *Clie
...
+		DeviceIdsFilter: []string{deviceID},
+	})
+	if err != nil {
+		if status.Convert(err).Code() == codes.NotFound {
+			return nil
+		}
+		return fmt.Errorf("cannot get device(%v) metdata: %v", deviceID, err)
+	}
+	shadowSynchronizationDisabled := false
+	for {
```
Comment (ondrejtomcik) [inline coap-gateway/service/signIn.go:79]: there will be only one response, why do you need for.

File: grpc-gateway/service/retrieveDevicesMetadata.go [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+		resp, err := rd.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return log.LogAndReturnError(kitNetGrpc.ForwardErrorf(codes.Internal, "cannot receive devices metadata: %v", err))
+		}
+		err = srv.Send(resp)
+		if err != nil {
+			return log.LogAndReturnError(kitNetGrpc.ForwardErrorf(codes.Internal, "cannot send devices metadata: %v", err))
```
Comment (ondrejtomcik) [inline grpc-gateway/service/retrieveDevicesMetadata.go]: cannot return fits better.

File: grpc-gateway/service/subscribeToEvents.go [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+
+	"google.golang.org/grpc/codes"
+)
+
+func (r *RequestHandler) SubscribeToEvents(srv pb.GrpcGateway_SubscribeToEventsServer) (errRet error) {
+	ctx, cancel := context.WithCancel(srv.Context())
+	defer cancel()
+	rd, err := r.resourceDirectoryClient.SubscribeToEvents(ctx)
+	if err != nil {
+		return log.LogAndReturnError(kitNetGrpc.ForwardErrorf(codes.Internal, "cannot subscribe for events: %v", err))
```
Comment (ondrejtomcik) [inline grpc-gateway/service/subscribeToEvents.go]: to events :) install gramarly

File: grpc-gateway/service/updateDeviceShadowSynchronization.go [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+				Disabled: req.GetDisabled(),
+			},
+		},
+		CommandMetadata: &commands.CommandMetadata{
+			ConnectionId: connectionID,
+		},
+	})
+
+	if err != nil {
+		return nil, log.LogAndReturnError(kitNetGrpc.ForwardErrorf(codes.Internal, "cannot update shadow synchronization of device %v: %v", req.GetDeviceId(), err))
```
Comment (ondrejtomcik) [inline grpc-gateway/service/updateDeviceShadowSynchronization.go:39]: %w?

File: resource-aggregate/commands/resourceconv.go [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+func SchemaResourceLinkToResource(link schema.ResourceLink, validUntil time.Time) *Resource {
+	return &Resource{
+		Href:                  link.Href,
+		DeviceId:              link.DeviceID,
+		ResourceTypes:         link.ResourceTypes,
+		Interfaces:            link.Interfaces,
+		Anchor:                link.Anchor,
+		Title:                 link.Title,
+		SupportedContentTypes: link.SupportedContentTypes,
+		ValidUntil:            validUntil.UnixNano(),
```
Comment (ondrejtomcik) [inline resource-aggregate/commands/resourceconv.go:93]: Aren't you loosing here former validUntil?

File: resource-aggregate/pb/resources.proto [hunk 22]
```diff
@@ -13,7 +13,7 @@ message Resource {
     string anchor = 5;
     string title = 6;
     repeated string supported_content_types = 7;
-    int32 time_to_live = 8;
+    int64 valid_until = 8;
```
Comment (ondrejtomcik) [inline resource-aggregate/pb/resources.proto:16]: why in64? what format is than valid until?

File: coap-gateway/service/resourceDirectory.go [hunk 23]
```diff
@@ -91,13 +90,9 @@ func resourceDirectoryPublishHandler(req *mux.Message, client *Client) {
...
 	}
 
-	for _, resource := range publishedResources {
-
-		err := client.observeResource(req.Context, resource.GetResourceID(), resource.IsObservable(), true)
-		if err != nil {
-			log.Errorf("unable to start observation of %v", resource.GetResourceID(), err)
-		}
-	}
+	client.server.taskQueue.Submit(func() {
```
Comment (ondrejtomcik) [inline coap-gateway/service/resourceDirectory.go:93]: is fixTTL func in resourceDirectory.go still required?

File: resource-aggregate/pb/resources.proto (hunk 22 above)
Comment (ondrejtomcik) [reply resource-aggregate/pb/resources.proto:16]: What if coap device retrieves RD? Do you calculate ttl back from validUntil-currentTime ? I cannot find this update.

File: resource-directory/service/deviceDirectory.go [hunk 24]
```diff
@@ -103,19 +101,21 @@ func updateDevice(dev *Device, resource *Resource) error {
...
 		dev.Resource.Interfaces = resource.Resource.GetInterfaces()
-	case cloudResourceTypes.HasOneOf(resource.Resource.GetResourceTypes()...):
-		var cloudStatus deviceStatus.Status
-		err := decodeContent(resource.GetContent(), &cloudStatus)
-		if err != nil {
-			return err
-		}
-		dev.IsOnline = cloudStatus.IsOnline()
-		dev.cloudStateUpdated = true
+		/*
```
Comment (ondrejtomcik) [inline resource-directory/service/deviceDirectory.go]: You forgot to finish this

File: resource-directory/service/resourceShadow.go [hunk 25]
```diff
@@ -150,7 +65,7 @@ func (rd *ResourceShadow) filterResources(ctx context.Context, resourceIDsFilter
 	return resources, err
 }
 
-func (rd *ResourceShadow) RetrieveResourcesValues(req *pb.RetrieveResourcesValuesRequest, srv pb.GrpcGateway_RetrieveResourcesValuesServer) error {
+func (rd *ResourceShadow) RetrieveResources(req *pb.RetrieveResourcesRequest, srv pb.GrpcGateway_RetrieveResourcesServer) error {
```
Comment (ondrejtomcik) [inline resource-directory/service/resourceShadow.go:69]: Where do you handle the situation that the device shadow sync was disabled and `retrieveResources` shall return unavailable?

**Comment by ondrejtomcik** [comment]
I would like to introduce configuration in RA shadowSynchronization.enabled and then block devicemetadataupdate commands enabling shadowsync in case enabled is set to false. Shortly, globally prevent usage of shadowsync.

**Comment by ondrejtomcik** [comment]
I don't like that it's shadowSync.disabled . We have everywhere enabled but here disabled. On how many places do you make use of default false value -> disabled : false? I don't think it's wort it. For client applications it's more common to work with positive boolean definitions.

**Review by ondrejtomcik — CHANGES_REQUESTED** [review]

**Comment by jkralik** [comment]
> 
> 
> I don't like that it's shadowSync.disabled . We have everywhere enabled but here disabled. On how many places do you make use of default false value -> disabled : false? I don't think it's wort it. For client applications it's more common to work with positive boolean definitions.

This is a problem with protobuf. proto v3 doesn't support default value and `shadowSync.disabled` is bool type. To solve comment I need to changed it to type enum.  Is it OK ?

File: coap-gateway/service/client.go (hunk 2 above)
Comment (jkralik) [reply coap-gateway/service/client.go]: `resourceObservationDisabled?`

File: coap-gateway/service/resourceDiscovery.go (hunk 9 above)
Comment (jkralik) [reply coap-gateway/service/resourceDiscovery.go:76]: you need to use 2 lines instead one:
- cast to type eg `type Resources []*Resource`
- and call function

File: coap-gateway/service/signIn.go (hunk 17 above)
Comment (jkralik) [reply coap-gateway/service/signIn.go:79]: there could be more pending updates - eg:
at 8am you set disable
at 10am you set enable
at 11am device establish connection to cloud ...

File: grpc-gateway/pb/devices.proto (hunk 14 above)
Comment (jkralik) [reply grpc-gateway/pb/devices.proto:120]: few lines below you can see them

File: grpc-gateway/pb/service.proto (hunk 16 above)
Comment (jkralik) [reply grpc-gateway/pb/service.proto:42]: Yes

File: grpc-gateway/service/subscribeToEvents.go (hunk 19 above)
Comment (jkralik) [reply grpc-gateway/service/subscribeToEvents.go]: grammarly :)

File: resource-aggregate/commands/resourceconv.go (hunk 21 above)
Comment (jkralik) [reply resource-aggregate/commands/resourceconv.go:93]: It converts time to unix timestamp in nano seconds.

File: resource-aggregate/pb/resources.proto (hunk 22 above)
Comment (jkralik) [reply resource-aggregate/pb/resources.proto:16]: `time.UnixNano` converts time to timestamp in nanoseconds. TTL is not part of `ResourceLink` schema.  It is part of resource publish schema so you don't provide the value, but you are right I don't filter out expired links.

**Comment by ondrejtomcik** [comment]
> This is a problem with protobuf. proto v3 doesn't support default value and shadowSync.disabled is bool type. To solve comment I need to changed it to type enum. Is it OK ?

As discussed, we will model it as enum, directly as a field included in the DeviceMetadataUpdated event.

File: cloud2cloud-connector/service/deviceSubscriptionHandlers.go (hunk 1 above)
Comment (jkralik) [reply cloud2cloud-connector/service/deviceSubscriptionHandlers.go]: fixed

File: coap-gateway/service/client.go (hunk 3 above)
Comment (jkralik) [reply coap-gateway/service/client.go]: fixed

File: coap-gateway/service/client.go (hunk 4 above)
Comment (jkralik) [reply coap-gateway/service/client.go]: fixed

File: coap-gateway/service/client.go (hunk 5 above)
Comment (jkralik) [reply coap-gateway/service/client.go]: fixed

File: coap-gateway/service/client.go (hunk 7 above)
Comment (jkralik) [reply coap-gateway/service/client.go]: fixed

File: coap-gateway/service/resourceDiscovery.go (hunk 10 above)
Comment (jkralik) [reply coap-gateway/service/resourceDiscovery.go]: fixed

File: grpc-gateway/service/retrieveDevicesMetadata.go (hunk 18 above)
Comment (jkralik) [reply grpc-gateway/service/retrieveDevicesMetadata.go]: fixed - added resp

File: grpc-gateway/service/updateDeviceShadowSynchronization.go (hunk 20 above)
Comment (jkralik) [reply grpc-gateway/service/updateDeviceShadowSynchronization.go:39]: `%w` no is not supported by grpc status.

File: coap-gateway/service/resourceDirectory.go (hunk 23 above)
Comment (jkralik) [reply coap-gateway/service/resourceDirectory.go:93]: we need to leave it is and plan to remove it when it will be fixed in device.
https://github.com/iotivity/iotivity-lite/blob/311705219fd4c17fb5dc858a386e051284c14a80/api/cloud/rd_client.c#L69

File: resource-directory/service/deviceDirectory.go (hunk 24 above)
Comment (jkralik) [reply resource-directory/service/deviceDirectory.go]: removed

File: resource-directory/service/resourceShadow.go (hunk 25 above)
Comment (jkralik) [reply resource-directory/service/resourceShadow.go:69]: We need to ask a customer.

File: grpc-gateway/service/subscribeToEvents.go (hunk 19 above)
Comment (jkralik) [reply grpc-gateway/service/subscribeToEvents.go]: fixed

**Comment by jkralik** [comment]
> 
> 
> > This is a problem with protobuf. proto v3 doesn't support default value and shadowSync.disabled is bool type. To solve comment I need to changed it to type enum. Is it OK ?
> 
> As discussed, we will model it as enum, directly as a field included in the DeviceMetadataUpdated event.

Fixed - shadowSync as enum

File: grpc-gateway/client/getDevices_test.go (hunk 12 above)
Comment (jkralik) [reply grpc-gateway/client/getDevices_test.go]: fixed

File: coap-gateway/service/client.go (hunk 8 above)
Comment (jkralik) [reply coap-gateway/service/client.go]: fixed

File: coap-gateway/service/client.go (hunk 6 above)
Comment (jkralik) [reply coap-gateway/service/client.go]: renamed to client.shadowSynchronization

File: coap-gateway/service/client.go (hunk 2 above)
Comment (jkralik) [reply coap-gateway/service/client.go]: renamed to shadowSynchronization

**Comment by jkralik** [comment]
> 
> 
> I don't like that it's shadowSync.disabled . We have everywhere enabled but here disabled. On how many places do you make use of default false value -> disabled : false? I don't think it's wort it. For client applications it's more common to work with positive boolean definitions.

 shadowSync.disabled was changed to enum shadowSynchronization

**Comment by jkralik** [comment]
> 
> 
> I would like to introduce configuration in RA shadowSynchronization.enabled and then block devicemetadataupdate commands enabling shadowsync in case enabled is set to false. Shortly, globally prevent usage of shadowsync.

We agreed that we drop the feature for now.

File: grpc-gateway/pb/retrieveDevicesMetadata.proto (hunk 15 above)
Comment (jkralik) [reply grpc-gateway/pb/retrieveDevicesMetadata.proto:11]: added

File: coap-gateway/service/resourceDirectory.go (hunk 23 above)
Comment (ondrejtomcik) [reply coap-gateway/service/resourceDirectory.go:93]: Are we going to fix it in iotivity-lite? Can you please create a bug in their github ?
Thanks

**Review by ondrejtomcik — APPROVED** [review]
